---
title: "SeaGen_RDA_GenEnv"
author: "Amy Zyck"
date: "3/21/2021"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Seascape Redundancy Analysis

This code follows that documented by [Tom Jenkins](https://github.com/Tom-Jenkins/seascape_rda_tutorial).

## [Prepare genetic data for redundancy analysis](https://github.com/Tom-Jenkins/seascape_rda_tutorial/blob/master/1.Prepare_genetic_data/1.prepare_genetic_data.R).

#### Notes before execution:
1. Make sure all required R packages are installed.
2. Set working directory to the location of this R script.

```{r}
# Load packages
library(adegenet)
library(poppr)
library(dplyr)
library(reshape2)
library(ggplot2)
library(vcfR)
```

# Outlier SNPs

## Making files

**Make genind object**

`outlierlocihap.recode.vcf` contains outlier SNPs (putatively under selection) for populations PVD, GB, BIS, and NAR called from reads aligned to the Eastern Oyster haplotig masked genome. Steps for generating this VCF file are located in [EecSeq_Cvirginica_dDocent.md](https://github.com/amyzyck/EecSeq_NB_EasternOyster/blob/master/Analysis/EecSeq_Cvirginica_dDocent.md) and [EecSeq_Cvirginica_Filtering.md](https://github.com/amyzyck/EecSeq_NB_EasternOyster/blob/master/Analysis/EecSeq_Cvirginica_Filtering.md). Population NIN was removed from the VCF file following steps in [EecSeq_Cvirginica_OutlierDetection.md](https://github.com/amyzyck/EecSeq_NB_EasternOyster/blob/master/Analysis/EecSeq_Cvirginica_OutlierDetection.md).

`strata` contains population, environmental, and library information for each sample - can be accessed [here](https://github.com/amyzyck/EecSeq_NB_EasternOyster/blob/master/Analysis/PopGen_SeaGen_Analyses/strata).

```{r}
my_vcf <- read.vcfR("outlierlocihap.recode.vcf")
strata <- read.table("strata", header=TRUE)
```

```{r}
rad.filt <- vcfR2genind(my_vcf, strata = strata, pop = c(rep("BIS", 10),rep("GB", 10),rep("NAR", 10), rep("PVD",10)))
```

```{r}
# Explore data
rad.filt
nLoc(rad.filt) # number of loci
nPop(rad.filt) # number of sites
nInd(rad.filt) # number of individuals
summary(rad.filt$pop) # sample size
```

```{r}
# Calculate allele frequencies for each site
allele_freqs = data.frame(rraf(rad.filt, by_pop = TRUE, correction = FALSE), check.names = FALSE)
```

```{r}
# Keep only the first of the two alleles for each SNP (since p=1-q).
allele_freqs = allele_freqs[, seq(1, dim(allele_freqs)[2], 2)]
```

```{r}
# Export allele frequencies
write.csv(allele_freqs, file = "all_allele_freqs.csv", row.names = TRUE)
```

#--------------#
#
# Calculate minor allele frequencies
#
#--------------#

```{r}
# Separate genind object by site
site_list = seppop(rad.filt)
names(site_list)
```

```{r}
# Calculate the minor allele frequency for each site
maf_list = lapply(site_list, FUN = minorAllele)
```

```{r}
# Convert list to dataframe
maf = as.data.frame(maf_list) %>% t() %>% as.data.frame()
```

```{r}
# Export minor allele frequencies
write.csv(maf, file = "minor_allele_freqs.csv", row.names = TRUE)
```

# ----------------- #
#
# Visualise allele frequencies
#
# ----------------- #

```{r}
# Add site labels
allele_freqs$site = rownames(allele_freqs)
```

```{r}
# Function to add regional labels to dataframe
addregion = function(x){
  # If pop label is present function will output the region
  if(x=="BIS") y = " Bissel Cove "
  if(x=="GB") y = " Greenwich Bay "
  if(x=="NAR") y = " Narrow River "
  if(x=="PVD") y = " Bold Point Park "
  return(y)
}
```

```{r}
# Add regional labels
allele_freqs$region = sapply(rownames(allele_freqs), addregion)
```

```{r}
# Convert dataframe to long format
allele_freqs.long = melt(allele_freqs, id.vars=c("site","region"))
```

```{r}
# Define order of facets using the levels argument in factor
unique(allele_freqs.long$site)
site_order =  c("BIS","GB","NAR","PVD")
allele_freqs.long$site_ord = factor(allele_freqs.long$site, levels = site_order)
```

```{r}
# Define region order
region_order = c(" Bissel Cove "," Greenwich Bay "," Narrow River ", " Bold Point Park ")
allele_freqs.long$region = factor(allele_freqs.long$region, levels = region_order)
```

```{r}
# Create colour scheme
# blue=#377EB8, green=#7FC97F, orange=#FDB462, red=#E31A1C
col_scheme = c("#7FC97F","#377EB8","#FDB462","#E31A1C")
```


A subset of the putatively outlier loci was selected, spanning all 4 outlier detection programs. 

```{r}
# Vector of outlier SNP loci to subset
desired_loci = c("NC_035780.1_5794934","NC_035780.1_17667463","NC_035780.1_57223575","NC_035781.1_38705306","NC_035781.1_52953122","NC_035782.1_40029780","NC_035783.1_27917169","NC_035784.1_24983069","NC_035784.1_79543049","NC_035785.1_32461323","NC_035787.1_28878662")
desired_loci_ID = sapply(paste(desired_loci, "..", sep = ""),
                         grep,
                         levels(allele_freqs.long$variable),
                         value = TRUE) %>% as.vector()
```


```{r}
# Subset dataset to plot desired SNP loci
allele_freqs.sub = allele_freqs.long %>% filter(variable %in% desired_loci_ID)
```

```{r}
# ggplot2 theme
ggtheme = theme(
  axis.text.x = element_blank(),
  axis.text.y = element_text(colour="black", size=6),
  axis.title = element_text(colour="black", size=15),
  panel.background = element_rect(fill="white"),
  panel.grid.minor = element_blank(),
  panel.grid.major = element_blank(),
  panel.border = element_rect(colour="black", fill=NA, size=0.5),
  plot.title = element_text(hjust = 0.5, size=18),
  legend.title = element_blank(),
  legend.text = element_text(size=10),
  legend.position = "top",
  legend.justification = "centre",
  # facet labels
  strip.text = element_text(colour="black", size=8)
)
```

```{r}
# Plot barplot
ggplot(data = allele_freqs.sub, aes(x = site_ord, y = value, fill = region))+
  geom_bar(stat = "identity", colour = "black", size = 0.3)+
  facet_wrap(~variable, scales = "free")+
  scale_y_continuous(limits = c(0,1), expand = c(0,0))+
  scale_fill_manual(values = col_scheme)+
  ylab("Allele frequency")+
  xlab("Site")+
  ggtheme
ggsave("allele_freq.png", width=10, height=8, dpi=300)
ggsave("allele_freq.pdf", width=10, height=8)
```


## [Prepare environmental data for redundancy analysis](https://github.com/Tom-Jenkins/seascape_rda_tutorial/blob/master/3.Prepare_environmental_data/3.prepare_env_data.R).

#### Environmental variables:
- Distance from sewage effluent source (km)
- Sewage Effluent (PW stats)
- Mean temperature (deg C)
- Mean Salinity (psu)
- Mean pH
- Mean Chlorophyll-a (ug/L)
- Mean Dissolved Oxygen (mg/L)

```{r}
# All environmental data was previously saved in strata file 
strata_pop <- read.table("strata_pop", header=TRUE)
strata_pop
```

```{r}
# Export data as a csv file
write.csv(strata_pop, file="environmental_data.csv", row.names = FALSE)
```

I also prepared spatial data for the redundancy analysis which is documented [here](https://github.com/amyzyck/EecSeq_NB_EasternOyster/blob/master/Analysis/PopGen_SeaGen_Analyses/RedundancyAnalysis/PrepSpatialData.Rmd).

Allele frequency, environmental, and spatial csv files are saved to your working directory and must be imported into the Rscript to run the redundancy analysis. Documentation of the RDA can be accessed [here](https://github.com/amyzyck/EecSeq_NB_EasternOyster/blob/master/Analysis/PopGen_SeaGen_Analyses/RedundancyAnalysis/RDA_Outlier_Hap.Rmd). 