---
title: "NB PopSea Gen Analysis Outlier SNPs for 6 pops"
author: "Amy Zyck"
date: '2023-03-16'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Population analyses on Outlier SNP dataset for 6 populations

I first created a new directory `OutlierSNPs_6` within `PATH: /home/azyck/NB_capture_both/NB_ddhaplo/NB_PopSeaGenAnalysis/`. I linked `outlierloci6.recode.vcf` and the working `strata_6pops` file (environmental and pop info for each individual) to this working directory. The RMarkdown file should be saved in this same working directory. 


```{r}
#Loading all the necessary packages
library(adegenet)
library(vcfR)
library("radiator") # Conversion from vcf to a lot of other formats
library("dplyr")
library("hierfstat")
library("ggplot2") #For plotting
library("reshape2") #For plotting
library("plyr")
library("cowplot") #For plotting manuscript figs
library(PCAviz)  #Visualizing output of PCA
library("stringr")
library("bigsnpr") # package for Linkage Disequilibrium pruning
```


# Outlier SNPs (6 populations)

## Making files

**Make genind object**

```{r}
my_vcf_out6 <- read.vcfR("outlierloci6.recode.vcf")
strata_out6 <- read.table("strata_6pops", header=TRUE)
```

```{r}
my_vcf_out6_nolfmm <- read.vcfR("6pops_outlierloci_nolfmm.recode.vcf")
strata_out6_nolfmm <- read.table("strata_6pops", header=TRUE)
```

```{r}
rad_out6.filt <- vcfR2genind(my_vcf_out6, strata = strata_out6, pop = c(rep("BAR", 10),rep("BIS", 10),rep("GB", 10),rep("KIC", 10),rep("MCD", 10), rep("PVD",10)))
```

```{r}
rad_out6.filt
```

```{r}
rad_out6_nolfmm.filt <- vcfR2genind(my_vcf_out6_nolfmm, strata = strata_out6, pop = c(rep("BAR", 10),rep("BIS", 10),rep("GB", 10),rep("KIC", 10),rep("MCD", 10), rep("PVD",10)))
```

```{r}
rad_out6_nolfmm.filt
```

```{r}
#Providing population names for plotting
pop_order_6pops <- c("BAR","BIS","GB","KIC","MCD","PVD")
```

**Read in the other info from .strata file and extract information such as locality, latitude, and longitude.**
```{r}
info_out6 <- as.data.frame(read.table("strata_6pops",header = T,sep = "\t",stringsAsFactors = F))
mystrats_out6 <- as.data.frame(matrix(nrow = length(indNames(rad_out6.filt)),ncol=9))
colnames(mystrats_out6) <- c("Population","Latitude","Longitude","Distance","SE","Temperature","Salinity","pH","DO")
just_out6.strats <- select(info_out6,c("Population"))
stratted_out6.filt <- strata(rad_out6.filt, formula= Population, combine = TRUE,just_out6.strats)
stratted_out6.filt@other <- select(info_out6,Latitude.,Longitude.,SewageEffluent,Temperature,Salinity,pH,DO.)
```

```{r}
stratted_out6.filt
```

```{r}
info_out6_nolfmm <- as.data.frame(read.table("strata_6pops",header = T,sep = "\t",stringsAsFactors = F))
mystrats_out6_nolfmm <- as.data.frame(matrix(nrow = length(indNames(rad_out6_nolfmm.filt)),ncol=9))
colnames(mystrats_out6_nolfmm) <- c("Population","Latitude","Longitude","Distance","SE","Temperature","Salinity","pH","DO")
just_out6_nolfmm.strats <- select(info_out6_nolfmm,c("Population"))
stratted_out6_nolfmm.filt <- strata(rad_out6_nolfmm.filt, formula= Population, combine = TRUE,just_out6_nolfmm.strats)
stratted_out6_nolfmm.filt@other <- select(info_out6_nolfmm,Latitude.,Longitude.,SewageEffluent,Temperature,Salinity,pH,DO.)
```

```{r}
stratted_out6_nolfmm.filt
```

**Make hierfstat object**
```{r}
hf_out6.filt <- genind2hierfstat(rad_out6.filt, pop = c(rep("BAR", 10),rep("BIS", 10),rep("GB", 10),rep("KIC", 10),rep("MCD", 10), rep("PVD",10)))
```

```{r}
hf_out6_nolfmm.filt <- genind2hierfstat(rad_out6_nolfmm.filt, pop = c(rep("BAR", 10),rep("BIS", 10),rep("GB", 10),rep("KIC", 10),rep("MCD", 10), rep("PVD",10)))
```

## Pairwise Fst

```{r}
fst_out6.mat <- pairwise.WCfst(hf_out6.filt)
```

```{r}
gindF_out6.fst.mat.triN <- as.matrix(fst_out6.mat)
colnames(gindF_out6.fst.mat.triN) <- pop_order_6pops
rownames(gindF_out6.fst.mat.triN) <- pop_order_6pops
```

```{r}
meltedN_out6 <- melt(gindF_out6.fst.mat.triN, na.rm =TRUE)
round(gindF_out6.fst.mat.triN,4)
```

```{r}
summary(meltedN_out6$value)
```


```{r}
#Plotting Pairwise fst
out6 <- ggplot(data = meltedN_out6, aes(Var2, Var1, fill = value))+ geom_tile(color = "white")+ 
  scale_fill_gradient(low = "white", high = "blue", name="FST")  +
  ggtitle(expression(atop("Pairwise FST, WC (1984) Outlier SNPs for 6 Populations", atop(italic("N = 60, L = 57"), ""))))+
  labs( x = "Sampling Site", y = "Sampling Site") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1),axis.text.y = element_text(size = 12)) + 
  theme(axis.title = element_text(size = 12),legend.text = element_text(size =14), legend.title = element_text(size =14)) +
  theme(plot.title = element_text(size = 14)) +
coord_fixed()
out6
```

## Pairwise Fst

```{r}
fst_out6_nolfmm.mat <- pairwise.WCfst(hf_out6_nolfmm.filt)
```

```{r}
gindF_out6_nolfmm.fst.mat.triN <- as.matrix(fst_out6_nolfmm.mat)
colnames(gindF_out6_nolfmm.fst.mat.triN) <- pop_order_6pops
rownames(gindF_out6_nolfmm.fst.mat.triN) <- pop_order_6pops
```

```{r}
meltedN_out6_nolfmm <- melt(gindF_out6_nolfmm.fst.mat.triN, na.rm =TRUE)
round(gindF_out6_nolfmm.fst.mat.triN,4)
```

```{r}
summary(meltedN_out6_nolfmm$value)
```


```{r}
#Plotting Pairwise fst
out6_nolfmm <- ggplot(data = meltedN_out6_nolfmm, aes(Var2, Var1, fill = value))+ geom_tile(color = "white")+ 
  scale_fill_gradient(low = "white", high = "blue", name="FST")  +
  ggtitle(expression(atop("Pairwise FST, WC (1984) Outlier SNPs for 6 Populations", atop(italic("N = 60, L = 18"), ""))))+
  labs( x = "Sampling Site", y = "Sampling Site") + 
  theme(axis.text.x = element_text(angle = 45, vjust = 1, size = 12, hjust = 1),axis.text.y = element_text(size = 12)) + 
  theme(axis.title = element_text(size = 12),legend.text = element_text(size =14), legend.title = element_text(size =14)) +
  theme(plot.title = element_text(size = 14)) +
coord_fixed()
out6_nolfmm
```


# Seascape Redundancy Analysis

This code follows that documented by [Tom Jenkins](https://github.com/Tom-Jenkins/seascape_rda_tutorial).

## [Prepare genetic data for redundancy analysis](https://github.com/Tom-Jenkins/seascape_rda_tutorial/blob/master/1.Prepare_genetic_data/1.prepare_genetic_data.R).

#### Notes before execution:
1. Make sure all required R packages are installed.
2. Set working directory to the location of this R script.

```{r}
# Load packages
library(adegenet)
library(poppr)
library(dplyr)
library(reshape2)
library(ggplot2)
library(vcfR)
```

```{r}
# Explore data
rad_out6.filt
nLoc(rad_out6.filt) # number of loci
nPop(rad_out6.filt) # number of sites
nInd(rad_out6.filt) # number of individuals
summary(rad_out6.filt$pop) # sample size
```

```{r}
# Explore data -no LFMM
rad_out6_nolfmm.filt
nLoc(rad_out6_nolfmm.filt) # number of loci
nPop(rad_out6_nolfmm.filt) # number of sites
nInd(rad_out6_nolfmm.filt) # number of individuals
summary(rad_out6_nolfmm.filt$pop) # sample size
```

```{r}
# Calculate allele frequencies for each site
allele_freqs_out6 = data.frame(rraf(rad_out6.filt, by_pop = TRUE, correction = FALSE), check.names = FALSE)
```

```{r}
# Calculate allele frequencies for each site - no LFMM
allele_freqs_out6_nolfmm = data.frame(rraf(rad_out6_nolfmm.filt, by_pop = TRUE, correction = FALSE), check.names = FALSE)
```

```{r}
# Keep only the first of the two alleles for each SNP (since p=1-q).
allele_freqs_out6 = allele_freqs_out6[, seq(1, dim(allele_freqs_out6)[2], 2)]
```

```{r}
# Keep only the first of the two alleles for each SNP (since p=1-q). - no LFMM
allele_freqs_out6_nolfmm = allele_freqs_out6_nolfmm[, seq(1, dim(allele_freqs_out6_nolfmm)[2], 2)]
```

```{r}
# Export allele frequencies
write.csv(allele_freqs_out6, file = "all_allele_freqs_out6.csv", row.names = TRUE)
```

```{r}
# Export allele frequencies - no LFMM
write.csv(allele_freqs_out6_nolfmm, file = "all_allele_freqs_out6_nolfmm.csv", row.names = TRUE)
```

#--------------#
#
# Calculate minor allele frequencies
#
#--------------#

```{r}
# Separate genind object by site
site_list_out6 = seppop(rad_out6.filt)
names(site_list_out6)
```

```{r}
# Separate genind object by site - no LFMM
site_list_out6_nolfmm = seppop(rad_out6_nolfmm.filt)
names(site_list_out6_nolfmm)
```

```{r}
# Calculate the minor allele frequency for each site
maf_list_out6 = lapply(site_list_out6, FUN = minorAllele)
```

```{r}
# Calculate the minor allele frequency for each site - no LFMM
maf_list_out6_nolfmm = lapply(site_list_out6_nolfmm, FUN = minorAllele)
```

```{r}
# Convert list to dataframe
maf_out6 = as.data.frame(maf_list_out6) %>% t() %>% as.data.frame()
```

```{r}
# Convert list to dataframe - no LFMM
maf_out6_nolfmm = as.data.frame(maf_list_out6_nolfmm) %>% t() %>% as.data.frame()
```

```{r}
# Export minor allele frequencies
write.csv(maf_out6, file = "minor_allele_freqs_out6.csv", row.names = TRUE)
```

```{r}
# Export minor allele frequencies - no LFMM
write.csv(maf_out6_nolfmm, file = "minor_allele_freqs_out6_nolfmm.csv", row.names = TRUE)
```

## [Prepare environmental data for redundancy analysis](https://github.com/Tom-Jenkins/seascape_rda_tutorial/blob/master/3.Prepare_environmental_data/3.prepare_env_data.R).

#### Environmental variables:
- Sewage Effluent (PW stats)
- Mean temperature (deg C)
- Mean Salinity (psu)
- Mean pH
- Mean Dissolved Oxygen (mg/L)

Environmental data for each population is saved in a `strata_pop6` file that can be accessed [here](https://github.com/amyzyck/EecSeq_NB_EasternOyster/blob/master/Analysis/PopGen_SeaGen_Analyses/strata_pop)

```{r}
# All environmental data was previously saved in strata file 
strata_pop6 <- read.table("strata_6pops", header=TRUE)
strata_pop6
```

```{r}
# Export data as a csv file
write.csv(strata_pop, file="environmental_data.csv", row.names = FALSE)
```

I also prepared spatial data for the redundancy analysis which is documented [here](https://github.com/amyzyck/EecSeq_NB_EasternOyster/blob/master/Analysis/PopGen_SeaGen_Analyses/RedundancyAnalysis/PrepSpatialData.Rmd).

Allele frequency, environmental, and spatial csv files are saved to your working directory and must be imported into the Rscript to run the redundancy analysis. 

# Redundancy Analysis

```{r}
# Load packages
library(psych)
library(dplyr)
library(adespatial)
library(vegan)
```

```{r}
# Import genetic data
allele_freqs_out6 = read.csv("all_allele_freqs_out6.csv", row.names = 1, check.names = FALSE)
```

```{r}
# Import genetic data - no LFMM
allele_freqs_out6_nolfmm = read.csv("all_allele_freqs_out6_nolfmm.csv", row.names = 1, check.names = FALSE)
```

```{r}
# Import environmental data
env.raw_out6 = read.csv("environmental_data_6pops.csv", row.names = 1)
```

```{r}
# Import spatial data
dbmem.raw_out6 = read.csv("dbmems_6pops.csv")
```

```{r}
# Set seed
set.seed(123)
```


#--------------#
#
# Multicollinearity checks
#
#--------------#

```{r}
# Plot and run correlation test on environmental variables
pairs.panels(env.raw_out6, scale = TRUE)
```
```{r}
# Remove correlated variables
env.data_out6 = subset(env.raw_out6, select = -c(Latitude,Longitude))
pairs.panels(env.data_out6, scale = TRUE)
```

```{r}
#standardize the environmental data 
# Scale and center variables
env.z_out6 <- decostand(env.data_out6, method = "standardize")
# Variables are now centered around a mean of 0
round(apply(env.z_out6, 2, mean), 1)
# and scaled to have a standard deviation of 1
apply(env.z_out6, 2, sd)
```

```{r}
# Combine all environmental variables and dbmems - no LFMM
env.full_out6_nolfmm = cbind(env.z_out6, dbmem.raw_out6)
str(env.full_out6_nolfmm)
```

#--------------#
#
# Identify significant variables
#
#--------------#

```{r}
# Use forward selection to identify significant environmental variables with ordiR2step
# first we neeed to create a null model and then a full model 
## Null model
RDA0_out6_nolfmm <- rda(allele_freqs_out6_nolfmm ~ 1,  env.full_out6_nolfmm) 

## Full model
RDAfull_out6_nolfmm <- rda(allele_freqs_out6_nolfmm ~ SewageEffluent + Temperature + Salinity + pH + DO + MEM1 + MEM2, env.full_out6_nolfmm)
```

```{r}
adjR2.RDAfull_out6_nolfmm <- RsquareAdj(RDAfull_out6_nolfmm)$adj.r.squared
```

```{r} 
# Running ordiR2step to identify significant environmental variables in the model 
mod <- ordiR2step(RDA0_out6_nolfmm, scope = formula(RDAfull_out6_nolfmm), Pin = 0.1, permutations = 1000, R2scope = FALSE, direction = "forward")
```

pH and Salinity identified as significant variables. I couldn't run ordiR2step with R2scope as true. This parameter uses adjusted R2 as the stopping criterion - only models with lower adjusted R2 than the scope are accepted. 



#--------------#
#
# Redundancy analysis
#
#--------------#

```{r}
# Perform RDA with all variables
rda1_out6 = rda(allele_freqs_out6 ~ ., data = env.dbmems_out6, scale = TRUE)
rda1_out6
```

```{r}
# Model summaries
RsquareAdj(rda1_out6) # adjusted Rsquared 
vif.cca(rda1_out6) # variance inflation factor (<10 OK)
anova.cca(rda1_out6, permutations = 1000) # full model
anova.cca(rda1_out6, permutations = 1000, by="margin") # per variable 
```

```{r}
# Variance explained by each canonical axis
summary(eigenvals(rda1_out6, model = "constrained"))
screeplot(rda1_out6)
```


```{r}
# Perform RDA with all variables - no LFMM
rda1_out6_nolfmm = rda(allele_freqs_out6_nolfmm ~ Salinity + pH, data = env.full_out6_nolfmm, scale = TRUE)
rda1_out6_nolfmm
```

```{r}
# Model summaries - no LFMM
RsquareAdj(rda1_out6_nolfmm) # adjusted Rsquared 
vif.cca(rda1_out6_nolfmm) # variance inflation factor (<10 OK)
anova.cca(rda1_out6_nolfmm, permutations = 1000) # full model
anova.cca(rda1_out6_nolfmm, permutations = 1000, by="margin") # per variable 
```

```{r}
# Variance explained by each canonical axis - no LFMM
summary(eigenvals(rda1_out6_nolfmm, model = "constrained"))
screeplot(rda1_out6_nolfmm)
```

```{r}
# Create a dataframe to correctly colour regions
col_dframe_out6 = data.frame("site" = rownames(allele_freqs_out6))
```

```{r}
# Create a dataframe to correctly colour regions - no LFMM
col_dframe_out6_nolfmm = data.frame("site" = rownames(allele_freqs_out6_nolfmm))
```

```{r}
# Function to add regional labels to dataframe
addregion_6pops = function(x){
  # If pop label is present function will output the region
  if(x=="BAR") y = " Barrington "
  if(x=="BIS") y = " Bissel Cove "
  if(x=="GB") y = " Greenwich Bay "
  if(x=="KIC") y = " Kickemuit "
  if(x=="MCD") y = " Donovan Marsh "
  if(x=="PVD") y = " Providence "
  return(y)
}
```

```{r}
# Add regional labels
col_dframe_out6$region = sapply(col_dframe_out6$site, addregion_6pops)
```

```{r}
# Add regional labels - no LFMM
col_dframe_out6_nolfmm$region = sapply(col_dframe_out6_nolfmm$site, addregion_6pops)
```

```{r}
# Add factor levels
region_order_6pops = c(" Barrington "," Bissel Cove "," Greenwich Bay "," Kickemuit "," Donovan Marsh ", " Providence ")
col_dframe_out6$region = factor(col_dframe_out6$region, levels = region_order_6pops)
```

```{r}
# Add factor levels - no LFMM
region_order_6pops = c(" Barrington "," Bissel Cove "," Greenwich Bay "," Kickemuit "," Donovan Marsh ", " Providence ")
col_dframe_out6_nolfmm$region = factor(col_dframe_out6_nolfmm$region, levels = region_order_6pops)
```

```{r}
# Create colour scheme
# blue=#000088, green=#7FC97F, orange=#FF7F00, red=#E31A1C, purple=#9A32CD, pink=#FF1493, yellow=#FFD700, cyan=#00FFFF
cols_6pops = c("#7FC97F","#00008B","#FF7F00","#9A32CD","#FF1493","#00FFFF")
```

```{r}
# Visualise results of RDA
png("rda_out6_nolegend.png", width = 8, height = 7, units = "in", res = 600)
plot(rda1_out6, type="n", scaling = 3)
# SITES
points(rda1_out6, display="sites", pch=21, scaling=3, cex=1.5, col="black",
       bg=cols_6pops[col_dframe_out6$region]) # sites
text(rda1_out6, display="sites", scaling = 3, col="black", font=2, pos=4)
# PREDICTORS
text(rda1_out6, display="bp", scaling=3, col="red1", cex=1, lwd=2)
# SNPS
# text(rda1, display="species", scaling = 3, col="blue", cex=0.7, pos=4) # SNPs
# LEGEND
#legend("bottomright", legend=levels(col_dframe_out6$region), bty="n", col="black",
      #pch=21, cex=1.2, pt.bg=cols_6pops)
# OTHER LABELS
adj.R2 = round(RsquareAdj(rda1_out6)$adj.r.squared, 3)
mtext(bquote(italic("R")^"2"~"= "~.(adj.R2)), side = 3, adj = 0.5)
dev.off()
```

```{r}
# Visualise results of RDA
png("rda_out6_nolfmm_pHSal_nopred.png", width = 8, height = 7, units = "in", res = 600)
plot(rda1_out6_nolfmm, type="n", scaling = 3)
# SITES
points(rda1_out6_nolfmm, display="sites", pch=21, scaling=3, cex=1.5, col="black",
       bg=cols_6pops[col_dframe_out6_nolfmm$region]) # sites
text(rda1_out6_nolfmm, display="sites", scaling = 3, col="black", font=2, pos=4)
# PREDICTORS
#text(rda1_out6_nolfmm, display="bp", scaling=3, col="red1", cex=1, lwd=2)
# SNPS
# text(rda1, display="species", scaling = 3, col="blue", cex=0.7, pos=4) # SNPs
# LEGEND
#legend("bottomright", legend=levels(col_dframe_out6$region), bty="n", col="black",
      #pch=21, cex=1.2, pt.bg=cols_6pops)
# OTHER LABELS
adj.R2 = round(RsquareAdj(rda1_out6_nolfmm)$adj.r.squared, 3)
mtext(bquote(italic("R")^"2"~"= "~.(adj.R2)), side = 3, adj = 0.5)
dev.off()
```

#--------------#
#
# Partial redundancy analysis
#
#--------------#

```{r}
# Perform RDA while controlling for geographical location 
pRDA_out6 = rda(allele_freqs_out6 ~ pH + Temperature + DO + Salinity + Condition(MEM2+MEM1),
           data = env.dbmems_out6, scale = TRUE)
pRDA_out6
RsquareAdj(pRDA_out6) # adjusted Rsquared 
vif.cca(pRDA_out6) # variance inflation factor (<10 OK)
anova.cca(pRDA_out6, permutations = 1000) # full model
anova.cca(pRDA_out6, permutations = 1000, by = "margin") # per variable
```



```{r}
# Perform RDA while controlling for geographical location - no LFMM
pRDA_out6_nolfmm = rda(allele_freqs_out6_nolfmm ~ Salinity + pH + Condition(MEM1+MEM2),
           data = env.full_out6_nolfmm, scale = TRUE)
pRDA_out6_nolfmm
RsquareAdj(pRDA_out6_nolfmm) # adjusted Rsquared 
vif.cca(pRDA_out6_nolfmm) # variance inflation factor (<10 OK)
anova.cca(pRDA_out6_nolfmm, permutations = 1000) # full model
anova.cca(pRDA_out6_nolfmm, permutations = 1000, by = "margin") # per variable
```



```{r}
# Variance explained by each canonical axis
summary(eigenvals(pRDA_out6, model = "constrained"))
screeplot(pRDA_out6)
```

```{r}
# Variance explained by each canonical axis - no LFMM
summary(eigenvals(pRDA_out6_nolfmm, model = "constrained"))
screeplot(pRDA_out6_nolfmm)
```

```{r}
# Visualise results of pRDA
png("partial_rda_out6_nopredictors.png", width = 9, height = 7, units = "in", res = 600)
plot(pRDA_out6, type="n", scaling = 3, xlab="57%", ylab= "23%")
#title("Seascape partial redundancy analysis")
# SITES
points(pRDA_out6, display="sites", pch=21, scaling=3, cex=1.5, col="black",
       bg=cols_6pops[col_dframe_out6$region]) # sites
text(pRDA_out6, display="sites", scaling = 3, col="black", font=2, pos=4)
# PREDICTORS
#text(pRDA_out6, display="bp", scaling=3, col="red1", cex=1, lwd=2)
# SNPS
# text(pRDA, display="species", scaling = 3, col="blue", cex=0.7, pos=4) # SNPs
# LEGEND
#legend("topleft", legend=levels(col_dframe_out6$region), bty="n", col="black",
       #pch=21, cex=1.2, pt.bg=cols_6pops)
# OTHER LABELS
#adj.R2 = round(RsquareAdj(pRDA)$r.squared, 3)
#mtext(bquote(italic("R")^"2"~"= "~.(adj.R2)), side = 3, adj = 0.5)
dev.off()
```

```{r}
# Visualise results of pRDA - no LFMM
png("partial_rda_out6_nolfmm_pHSal_largetxt.png", width = 9, height = 7, units = "in", res = 600)
plot(pRDA_out6_nolfmm, type="n", scaling = 3)
#title("Seascape partial redundancy analysis")
# SITES
points(pRDA_out6_nolfmm, display="sites", pch=21, scaling=3, cex=1.5, col="black",
       bg=cols_6pops[col_dframe_out6_nolfmm$region]) # sites
text(pRDA_out6_nolfmm, display="sites", scaling = 3, col="black", font=4, pos=3)
# PREDICTORS
text(pRDA_out6_nolfmm, display="bp", scaling=3, col="red1", cex=1.5, lwd=2)
# SNPS
# text(pRDA, display="species", scaling = 3, col="blue", cex=0.7, pos=4) # SNPs
# LEGEND
#legend("topleft", legend=levels(col_dframe_out6$region), bty="n", col="black",
       #pch=21, cex=1.2, pt.bg=cols_6pops)
# OTHER LABELS
adj.R2 = round(RsquareAdj(pRDA_out6_nolfmm)$adj.r.squared, 3)
mtext(bquote(italic("R")^"2"~"= "~.(adj.R2)), side = 3, adj = 0.5)
dev.off()
```

## Genetic diversity (observed and expected heterozygosity)

```{r}
comb_out6 <- summary(stratted_out6.filt)
names(comb_out6)
```

```{r}
plot(comb_out6$Hobs, xlab="Loci number", ylab="Observed Heterozygosity", 
     main="Observed heterozygosity per locus")
```

```{r}
plot(comb_out6$Hobs,comb_out6$Hexp, xlab="Hobs", ylab="Hexp", 
     main="Expected heterozygosity as a function of observed heterozygosity per locus")
```

```{r}
bartlett.test(list(comb_out6$Hexp, comb_out6$Hobs)) # a test : H0: Hexp = Hobs
```

**_Not a significant difference between Observed and expected heterzygosity_**.  

```{r}
basicstat_out6 <- basic.stats(hf_out6.filt, diploid = TRUE, digits = 3)
```

```{r}
as.data.frame(basicstat_out6$overall)
```


```{r}
# get bootstrap confidence values for Fis
boot_out6 <- boot.ppfis(hf_out6.filt,nboot = 1000)
boot5_out6 <- boot.ppfis(hf_out6.filt,nboot = 1000,quant = 0.5)
```

```{r}
# add latitude for each population
latitude_6pops = c(41.741, 41.545, 41.654, 41.698, 41.547, 41.816)
```

```{r}
# add longitude for each population
longitude_6pops = c(-71.305, -71.431, -71.445, -71.247, -71.203, -71.391)
```


```{r}
# add sewage effluent for each population
sewage_6pops = c(17.881, 8.825, 14.596, 56.313, 12.111, 59.860)
```

```{r}
# add temperature for each population
temperature_6pops = c(22.08, 21.39, 22.27, 21.50, 22.24, 15.80)
```

```{r}
# add salinity for each population
salinity_6pops = c(29.08, 27.32, 29.58, 28.31, 20.68, 18.82)
```

```{r}
# add pH for each population
pH_6pops = c(7.69, 7.94, 7.67, 7.84, 7.69, 7.68)
```

```{r}
# add DO for each population
DO_6pops = c(5.37, 7.05, 4.57, 6.07, 8.76, 4.90)
```

```{r}
# combine all pop statistics
colnames(basicstat_out6$Ho) <- pop_order_6pops
Ho_out6 <- colMeans(basicstat_out6$Ho,na.rm = T)
He_out6 <- colMeans(basicstat_out6$Hs,na.rm = T)
Fis_out6<- boot5_out6$fis.ci$ll
y_out6 <- cbind(pop_order_6pops,Ho_out6,He_out6,Fis_out6,boot_out6$fis.ci,latitude_6pops, longitude_6pops, sewage_6pops, temperature_6pops, salinity_6pops, pH_6pops, DO_6pops)
y_out6
```

```{r}
summary(He_out6)
```

```{r}
summary(Fis_out6)
```

```{r}
# Plot He vs Latitude
R2_out6 = round(summary(lm(y_out6$He_out6 ~ y_out6$latitude_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = latitude_6pops, y = He_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(15,18,17,16,6,3,4,9)) + 
geom_smooth(method=lm) + 
ggtitle("Expected heterozygosity vs Latitude, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=41.7, y=0.36, parse=T) +
scale_x_continuous()
```

```{r}
# Plot Ho vs Latitude
R2_out6 = round(summary(lm(y_out6$Ho_out6 ~ y_out6$latitude_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = latitude_6pops, y = Ho_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(15,18,17,16,6,3,4,9)) + 
geom_smooth(method=lm) + 
ggtitle("Observed heterozygosity vs Latitude, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=41.7, y=0.36, parse=T) +
scale_x_continuous()
```

```{r}
# Plot He vs Longitude
R2_out6 = round(summary(lm(y_out6$He_out6 ~ y_out6$longitude_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = longitude_6pops, y = He_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(15,18,17,16,6,3,4,9)) + 
geom_smooth(method=lm) + 
ggtitle("Expected heterozygosity vs Longitude, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=-71.3, y=0.36, parse=T) +
scale_x_continuous()
```
```{r}
# Plot Ho vs Longitude
R2_out6 = round(summary(lm(y_out6$Ho_out6 ~ y_out6$longitude_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = longitude_6pops, y = Ho_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(15,18,17,16,6,3,4,9)) + 
geom_smooth(method=lm) + 
ggtitle("Observed heterozygosity vs Longitude, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=-71.3, y=0.36, parse=T) +
scale_x_continuous()
```

```{r}
# Plot He vs Sewage Effluent
R2_out6 = round(summary(lm(y_out6$He_out6 ~ y_out6$sewage_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = sewage_6pops, y = He_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(15,18,17,16,6,3,4,9)) + 
geom_smooth(method=lm) + 
ggtitle("Expected heterozygosity vs Sewage Effluent, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=30, y=0.36, parse=T) +
scale_x_continuous()
```

```{r}
# Plot Ho vs Sewage Effluent
R2_out6 = round(summary(lm(y_out6$Ho_out6 ~ y_out6$sewage_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = sewage_6pops, y = Ho_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(15,18,17,16,6,3,4,9)) + 
geom_smooth(method=lm) + 
ggtitle("Observed heterozygosity vs Sewage Effluent, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=30, y=0.36, parse=T) +
scale_x_continuous()
```

```{r}
# Plot He vs Temperature
R2_out6 = round(summary(lm(y_out6$He_out6 ~ y_out6$temperature_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = temperature_6pops, y = He_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(15,18,17,16,6,3,4,9)) + 
geom_smooth(method=lm) + 
ggtitle("Expected heterozygosity vs Temperature, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=19, y=0.36, parse=T) +
scale_x_continuous()
```

```{r}
# Plot Ho vs Temperature
R2_out6 = round(summary(lm(y_out6$Ho_out6 ~ y_out6$temperature_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = temperature_6pops, y = Ho_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(15,18,17,16,6,3,4,9)) + 
geom_smooth(method=lm) + 
ggtitle("Observed heterozygosity vs Temperature, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=19, y=0.36, parse=T) +
scale_x_continuous()
```

```{r}
# Plot He vs Salinity
R2_out6 = round(summary(lm(y_out6$He_out6 ~ y_out6$salinity_6pops))$r.squared, 8)             
pHplot_out6 <-ggplot(y_out6, aes(x = salinity_6pops, y = He_out6)) + geom_point(aes(color=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_color_manual(values = c("#7FC97F","#00008B","#FF7F00","#9A32CD","#FF1493","#00FFFF")) + 
geom_smooth(method=lm) + 
# ggtitle("Expected heterozygosity vs Salinity, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=21, y=0.4, parse=T) +
labs(x = "Salinity", y = "Expected Heterozygosity") +
scale_x_continuous() +
theme(legend.title = element_blank()) 

pHplot_out6 + theme_classic()
```

```{r}
# Plot Ho vs Salinity
R2_out6 = round(summary(lm(y_out6$Ho_out6 ~ y_out6$salinity_6pops))$r.squared, 8)             
pHplot_out6 <-ggplot(y_out6, aes(x = salinity_6pops, y = Ho_out6)) + geom_point(aes(color=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_color_manual(values = c("#7FC97F","#00008B","#FF7F00","#9A32CD","#FF1493","#00FFFF")) + 
geom_smooth(method=lm) + 
# ggtitle("Observed heterozygosity vs Salinity, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=21, y=0.39, parse=T) +
labs(x = "Salinity", y = "Observed Heterozygosity") +
scale_x_continuous() +
theme(legend.title = element_blank()) 

pHplot_out6 + theme_classic()
```

```{r}
# Plot He vs pH
R2_out6 = round(summary(lm(y_out6$He_out6 ~ y_out6$pH_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = pH_6pops, y = He_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(15,18,17,16,6,3,4,9)) + 
geom_smooth(method=lm) + 
ggtitle("Expected heterozygosity vs pH, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=7.8, y=0.36, parse=T) +
scale_x_continuous()
```

```{r}
# Plot Ho vs pH
R2_out6 = round(summary(lm(y_out6$Ho_out6 ~ y_out6$pH_6pops))$r.squared, 8)             
pHplot_out6 <-ggplot(y_out6, aes(x = pH_6pops, y = Ho_out6)) + geom_point(aes(color=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_color_manual(values = c("#7FC97F","#00008B","#FF7F00","#9A32CD","#FF1493","#00FFFF")) + 
geom_smooth(method=lm) + 
# ggtitle("Observed heterozygosity vs pH, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=7.7, y=0.38, parse=T) +
labs(x = "pH", y = "Observed Heterozygosity") +
scale_x_continuous() +
theme(legend.title = element_blank()) 

pHplot_out6 + theme_classic()
```

```{r}
# Plot He vs DO
R2_out6 = round(summary(lm(y_out6$He_out6 ~ y_out6$DO_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = DO_6pops, y = He_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(15,18,17,16,6,3,4,9)) + 
geom_smooth(method=lm) + 
ggtitle("Expected heterozygosity vs Dissolved Oxygen, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=5.4, y=0.36, parse=T) +
scale_x_continuous()
```

```{r}
# Plot Ho vs DO
R2_out6 = round(summary(lm(y_out6$Ho_out6 ~ y_out6$DO_6pops))$r.squared, 8)             
ggplot(y_out6, aes(x = DO_6pops, y = Ho_out6)) + geom_point(aes(shape=pop_order_6pops, color=pop_order_6pops), size = 3) + scale_shape_manual(values = c(0,1,2,3,4,5)) + 
geom_smooth(method=lm) + 
ggtitle("Observed heterozygosity vs Dissolved Oxygen, Outlier SNPs 6 pops") +
annotate(geom = "text", label=paste("italic(R^2)==",R2_out6), x=5.4, y=0.36, parse=T) +
scale_x_continuous()
```

PCA 

```{r}
# Convert VCF to genotype matrix
geno_out6 <- extract.gt(my_vcf_out6) # Character matrix containing the genotypes
position_out6 <- getPOS(my_vcf) # Positions in bp
chromosome_out6 <- getCHROM(my_vcf) # Chromosome information

G_out6 <- matrix(NA, nrow = nrow(geno_out6), ncol = ncol(geno_out6))

G_out6[geno_out6 %in% c("0/0", "0|0")] <- 0
G_out6[geno_out6  %in% c("0/1", "1/0", "1|0", "0|1")] <- 1
G_out6[geno_out6 %in% c("1/1", "1|1")] <- 2

# NA should be replaced with “9” to work with the functions in the OutFLANK package
G_out6[is.na(G_out6)] <- 9

head(G_out6[,1:10])
```

```{r}
pca_out6 <- rda(G_out6)
```

```{r}
plot(pca_out6)
```


# Distance based RDA (dbRDA)

```{r}
rankindex(env.full_out6_nolfmm, allele_freqs_out6, indices = c("euc", "man", "gow","bra", "kul"), stepacross= FALSE, method = "spearman")
```

Trying out a distance based RDA (dbRDA)

```{r}
dbRDA_out6_noflmm = capscale(allele_freqs_out6 ~ pH + Salinity + Condition(MEM1+MEM2), env.full_out6_nolfmm, dist="kul")
```

```{r}
plot(dbRDA_out6_noflmm) # use base plot, might be done with ggplot2
anova(dbRDA_out6_noflmm) # is the model significant?
```

```{r}
anova(dbRDA_out6_noflmm) # overall test of the significant of the analysis
anova(dbRDA_out6_noflmm, by="axis", perm.max=1000) # test axes for significance
anova(dbRDA_out6_noflmm, by="terms", permu=1000) # test for sign. environ. variables
```


# PCA 

Following steps by 

```{r}
pca1_out6 <- read.table("plink2.eigenvec",sep="",header=F)
```
```{r}
eigenval_out6 <- scan("plink2.eigenval")
```

```{r}
plot(data=pca1_out6, V3~V4)
```

```{r}
# Creating list of each site multipled by number of individuals in site
pops_list6 <- c(rep("BAR", 10),rep("BIS", 10),rep("GB", 10),rep("KIC", 10),rep("MCD", 10), rep("PVD",10))
```

```{r}
pca_colors <- c(rep("#7FC97F",10),rep("#00008B",10),rep("#FF7F00",10),rep("#9A32CD",10),rep("#FF1493",10),rep("#00FFFF",10))
```

```{r}
#Adding population info to pca1 file
pca2_out6 <- cbind(pca1_out6, pops_list6)
```

```{r}
pca3_out3 <- cbind(pca2_out6,pca_colors)
```

```{r}
plot_colors <- pca3_out3$pca_colors
```

```{r}
# now working with the eigenvalue 
# first convert to percentage variance explained
pve <- data.frame(PC = 1:10, pve = eigenval_out6/sum(eigenval_out6)*100)
```

```{r}
# make plot of variance explained
a <- ggplot(pve, aes(PC, pve)) + geom_bar(stat = "identity")
a + ylab("Percentage variance explained") + theme_light()
```

```{r}
# calculate the cumulative sum of the percentage variance explained
cumsum(pve$pve)
```

```{r}
pca_out6 <- ggplot(data=pca2_out6, aes(V3,V4, color=pops_list6)) + geom_point(color = plot_colors, size = 3) + 
  ggtitle("All Outlier SNPs") + theme_light()
pca_out6 + xlab(paste0("PC1 (", signif(pve$pve[1], 3), "%)")) + ylab(paste0("PC2 (", signif(pve$pve[2], 3), "%)"))
```


Re running PCA on differentation based outliers only 

```{r}
pca1_out6dif <- read.table("diffout.eigenvec",sep="",header=F)
```
```{r}
eigenval_out6dif <- scan("diffout.eigenval")
```

```{r}
plot(data=pca1_out6dif, V3~V4)
```

```{r}
# Creating list of each site multipled by number of individuals in site
pops_list6 <- c(rep("BAR", 10),rep("BIS", 10),rep("GB", 10),rep("KIC", 10),rep("MCD", 10), rep("PVD",10))
```

```{r}
pca_colors <- c(rep("#7FC97F",10),rep("#00008B",10),rep("#FF7F00",10),rep("#9A32CD",10),rep("#FF1493",10),rep("#00FFFF",10))
```

```{r}
#Adding population info to pca1 file
pca2_out6dif <- cbind(pca1_out6dif, pops_list6)
```

```{r}
pca3_out6dif <- cbind(pca2_out6dif,pca_colors)
```

```{r}
plot_colorsdif <- pca3_out3dif$pca_colors
```

```{r}
# now working with the eigenvalue 
# first convert to percentage variance explained
pvedif <- data.frame(PC = 1:10, pve = eigenval_out6dif/sum(eigenval_out6dif)*100)
```

```{r}
# make plot of variance explained
adif <- ggplot(pvedif, aes(PC, pve)) + geom_bar(stat = "identity")
adif + ylab("Percentage variance explained") + theme_light()
```

```{r}
# calculate the cumulative sum of the percentage variance explained
cumsum(pvedif$pve)
```

```{r}
pca_out6dif <- ggplot(data=pca2_out6dif, aes(V3,V4, color=pops_list6)) + geom_point(color = plot_colors, size = 3) + 
  ggtitle("Differentation Outlier SNPs") + theme_light()
pca_out6dif + xlab(paste0("PC1 (", signif(pvedif$pve[1], 3), "%)")) + ylab(paste0("PC2 (", signif(pvedif$pve[2], 3), "%)"))
```

```{r}
X_out6 <- tab(stratted_out6.filt, freq = TRUE, NA.method = "mean")
pca1_out6 <- dudi.pca(X_out6, scale = FALSE, scannf = FALSE, nf = 3) 
barplot(pca1_out6$eig[1:50], main = "PCA eigenvalues", col = heat.colors(50))
```

```{r}
s.class(pca1_out6$li, pop(stratted_out6.filt))
title("PCA of neutral SNPs\naxes 1-2") 
add.scatter.eig(pca1_out6$eig[1:20], 3,1,2)
```

```{r}
pca_groups <- as.factor(pops_list6)
```

```{r}
#col <- funky(15)
s.class(pca1_out6$li,xax=1,yax=2, fac = pop(stratted_out6.filt), col= c("#7FC97F","#00008B","#FF7F00","#9A32CD","#FF1493","#00FFFF"), axesell=FALSE, cstar=0, cpoint=3, grid=FALSE)
```



```{r}
X_out6_nolfmm <- tab(stratted_out6_nolfmm.filt, freq = TRUE, NA.method = "mean")
pca1_out6_nolfmm <- dudi.pca(X_out6_nolfmm, scale = FALSE, scannf = FALSE, nf = 3) 
barplot(pca1_out6_nolfmm$eig[1:50], main = "PCA eigenvalues", col = heat.colors(50))
```

```{r}
s.class(pca1_out6_nolfmm$li, pop(stratted_out6_nolfmm.filt))
title("PCA of neutral SNPs\naxes 1-2") 
add.scatter.eig(pca1_out6_nolfmm$eig[1:16], 3,1,2)
```

```{r}
#col <- funky(15)
s.class(pca1_out6_nolfmm$li,xax=1,yax=2, fac = pop(stratted_out6_nolfmm.filt), col= c("#7FC97F","#00008B","#FF7F00","#9A32CD","#FF1493","#00FFFF"), axesell=FALSE, cstar=0, cpoint=3, grid=FALSE)
```
